include/dcgp/problems/symbolic_regression.hpp

```diff
@@ -1,19 +1,208 @@
+        std::transform(x.begin(), x.end(), xu.begin(), [](double a) { return boost::numeric_cast<unsigned>(a); });
+        m_cgp.set(xu);
+        std::ostringstream ss;
+        std::vector<std::string> symbols;
+        for (decltype(m_points[0].size()) i = 0u; i < m_points[0].size(); ++i) {
+            symbols.push_back("x" + std::to_string(i));
+        }
+        pagmo::stream(ss, m_cgp(symbols));
+        return ss.str();
+    }
+
+    /// Getter for the CGP
+    /**
+     * @return a const reference to the internal dcgp::expression<double> data member.
+     */
+    const expression<double> &get_cgp() const
+    {
+        return m_cgp;
+    }
+
+private:
+    inline void sanity_checks(unsigned &n, unsigned &m) const
+    {
+        // 1 - We check that points is not an empty vector.
+        if (m_points.size() == 0) {
+            throw std::invalid_argument("The size of the input data (points) is zero.");
+        }
+        n = static_cast<unsigned>(m_points[0].size());
+        m = static_cast<unsigned>(m_labels[0].size());
+        // 2 - We check labels and points have the same (non-empty) size
+        if (m_points.size() != m_labels.size()) {
+            throw std::invalid_argument("The number of input data (points) is " + std::to_string(m_points.size())
+                                        + " while the number of labels is " + std::to_string(m_labels.size())
+                                        + ". They should be equal.");
+        }
+        // 3 - We check that all p in points have the same size
+        if (!std::all_of(m_points.begin(), m_points.end(), [n](const std::vector<double> &p) { return p.size() == n; })) {
+            throw std::invalid_argument("The input data (points) is inconsistent: all points must have the same "
+                                        "dimension, while I detect differences.");
+        }
+        // 4 - We check that all l in labels have the same size
+        if (!std::all_of(m_labels.begin(), m_labels.end(), [m](const std::vector<double> &l) { return l.size() == m; })) {
+            throw std::invalid_argument("The labels are inconsistent: all labels must have the same "
+                                        "dimension, while I detect differences.");
+        }
+        if (m_c == 0) throw std::invalid_argument("Number of columns is 0");
+        if (m_r == 0) throw std::invalid_argument("Number of rows is 0");
+        if (m_l == 0) throw std::invalid_argument("Number of level-backs is 0");
+        if (m_arity < 2) throw std::invalid_argument("Arity must me at least 2.");
+        if (m_f.size() == 0) throw std::invalid_argument("Number of basis functions is 0");
+    }
+
+    std::vector<std::vector<double>> m_points;
+    std::vector<std::vector<double>> m_labels;
+    unsigned m_r;
+    unsigned m_c;
+    unsigned m_l;
+    unsigned m_arity;
+    std::vector<kernel<double>> m_f;
+    unsigned m_parallel_batches;
+    mutable expression<double> m_cgp;
```

Comment:
This has to be mutable because it's changed in the ``fitness()`` right? Note that this will hamper a bit the performance of ``bfe`` because the ``thread_bfe`` will make copies of the UDP for parallel evaluation. If this was not mutable, you could mark the UDP with a higher level of thread safety and you could avoid copies in the bfe.

Reply:
I cannot think of a way to avoid it, unfortunately the UDP contains the data whch can be large .... will think about a way pout, but it looks as necessary

Reply:
OK, the following would work:
Right now we have the following operator:
```
virtual std::vector<std::string> operator()(const std::vector<std::string> &point) const
```
this computes the symbolic expression encoded by **m_x**. SO before calling it one has to modify the object and set m_x. This updates data structures etc.

We could add:
```
std::vector<std::string> operator()(const std::vector<std::string> &point, std::vector<unsigned> x) const
```
this would give the same output avoiding to call a set_x before. We would have to check the sanity of the input **x**, but then we would be able to call it in the fitness of the UDP whithout having to modify the dcgp before.

Reply:
Perhaps we can keep it in mind as a future optimisation (i.e., put it in a comment somewhere)? Depending on how costly it is to call the ``fitness()`` vs copying the data structures in the ``bfe``, this might not turn out to be worthwhile pursuing. Or you already have a good idea on the relative costs of fitness vs copying?

Reply:
clearly not. will log it then and go on.